server/chat-plugins/blackjack.js

```diff
@@ -0,0 +1,643 @@
+
+		if (!force) {
+			if (this.dealer.points > 21) {
+				for (let player of Object.keys(this.playerTable)) {
+					if (this.playerTable[player].status === 'bust') continue;
+					winners.push(Chat.html`<strong>${this.playerTable[player].name}</strong> [${this.playerTable[player].cards.join(', ')}]`);
+				}
+			} else if (this.dealer.points !== 21) {
+				for (let player of Object.keys(this.playerTable)) {
+					if (this.playerTable[player].status === 'bust' || this.playerTable[player].points <= this.dealer.points) continue;
+					winners.push(Chat.html`<strong>${this.playerTable[player].name}</strong> [${this.playerTable[player].cards.join(', ')}]`);
+				}
+			} else if (this.dealer.points === 21) {
+				winners.push(`<strong>${this.dealer.name}</strong> [${this.dealer.cards.join(', ')}]`);
+			}
+			this.send(`[Blackjack has ended.]`, true);
+			if (winners.length < 1) {
+				this.display(`There are no winners this time.`, null, null, null, true);
+			} else {
+				this.display(`<strong>Winner${Chat.plural(winners.length)}</strong>: ${winners.join(', ')}`, null, null, null, true);
+			}
+		}
+
+		if (this.autostart) {
+			clearTimeout(this.autostart);
+			this.autostart = null;
+		}
+		this.state = 'ended';
+
+		this.destroy();
+	}
+	destroy() {
+		if (Object.keys(this.playerTable)) {
+			for (let player of Object.keys(this.playerTable)) {
+				this.playerTable[player].destroy();
+			}
+		}
+		this.room.blackjack++;
+		this.clearAllTimers();
+		delete this.room.game;
+	}
+
+	/**
+	 * Gameplay
+	 * hit - player decides to get a new card
+	 * stand - player decides to keep current cards
+	 * giveCard - gives a player a card from the deck
+	 * getCardPoints - returns the point value of a user's cards
+	 * next - game goes on to the next turn
+	 */
+	hit(user) {
+		if (this.state !== 'started') return this.errorMessage(user, `Blackjack hasn't started yet.`);
+		if (!this.playerTable[user]) return this.errorMessage(user, `You aren't a player in this game.`);
+		if (this.curUser !== user.userid) return this.errorMessage(user, `It's not your turn.`);
+		this.playerTable[user].selfUhtml = 'change';
+		this.playerTable[user].resetTimerTicks();
+
+		this.giveCard(user);
+	}
+	stand(user) {
+		const player = this.playerTable[user];
+		if (this.state !== 'started') return this.errorMessage(user, `Blackjack hasn't started yet.`);
+		if (!player) return this.errorMessage(user, `You aren't a player in this game.`);
+		if (this.curUser !== user.userid) return this.errorMessage(user, `It's not your turn.`);
+		player.status = 'stand';
+		let cards = '';
+		for (let card of player.cards) cards += `[${card}] `;
+		let turnLine = Chat.html`<br /><strong>${player.name}</strong> stands with ${cards} (${player.points}${player.points === 21 ? ' - blackjack!' : ''})`;
+
+		this.turnLog += turnLine;
+		this.display(turnLine, null, player.name);
+		this.clear();
+
+		this.next();
+	}
+	giveCard(user) {
+		if (this.deck.length < 1) this.deck = Dex.shuffle(this.deck);
```

Comment:
this appears to be `BlackjackDeck#length`, which doesn't exist

Reply:
No, ``this.deck`` here is assigned to ``BlackjackDeck#shuffle`` - which returns an array of the cards in a random order.

```Javascript
class BlackjackDeck {
	constructor() {
		this.deck = [
			'A♥', 'A♦', 'A♣', 'A♠', '2♥', '2♦', '2♣', '2♠', '3♥', '3♦', '3♣',
			'3♠', '4♥', '4♦', '4♣', '4♠', '5♥', '5♦', '5♣', '5♠', '6♥', '6♦', '6♣', '6♠',
			'7♥', '7♦', '7♣', '7♠', '8♥', '8♦', '8♣', '8♠', '9♥', '9♦', '9♣', '9♠', '10♥',
			'10♦', '10♣', '10♠', 'J♥', 'J♦', 'J♣', 'J♠', 'Q♥', 'Q♦', 'Q♣', 'Q♠', 'K♥', 'K♦',
			'K♣', 'K♠',
		];
	}
	shuffle() {
		return Dex.shuffle(this.deck);
	}
}
```

Reply:
My bad.

Would BlackjackDeck be better suited as a factory function instead, then?

Reply:
I'm not sure - it's relative to the specific game, which is why making it a class was my first instinct; it could just be called by ``new class`` which seemed easier at the time to manipulate.

Reply:
Semi-unrelated: this line was actually not generating a new deck right; I've just fixed that